src/decode.rs

```diff
@@ -3806,1159 +3806,204 @@ unsafe fn decode_b(
-                }
-                32 => {
-                    let const_val_86: uint64_t = (0x101010101010101 as libc::c_ulonglong)
-                        .wrapping_mul(b.c2rust_unnamed.c2rust_unnamed.uv_mode as libc::c_ulonglong)
-                        as uint64_t;
-                    (*(&mut *((*t.a).uvmode).as_mut_ptr().offset((cbx4 + 0) as isize)
-                        as *mut uint8_t as *mut alias64))
-                        .u64_0 = const_val_86;
-                    (*(&mut *((*t.a).uvmode).as_mut_ptr().offset((cbx4 + 8) as isize)
-                        as *mut uint8_t as *mut alias64))
-                        .u64_0 = const_val_86;
-                    (*(&mut *((*t.a).uvmode).as_mut_ptr().offset((cbx4 + 16) as isize)
-                        as *mut uint8_t as *mut alias64))
-                        .u64_0 = const_val_86;
-                    (*(&mut *((*t.a).uvmode).as_mut_ptr().offset((cbx4 + 24) as isize)
-                        as *mut uint8_t as *mut alias64))
-                        .u64_0 = const_val_86;
-                }
-                _ => {}
-            }
-            if b.c2rust_unnamed.c2rust_unnamed.pal_sz[1] != 0 {
-                let pal_0: *const [uint16_t; 8] = (if t.frame_thread.pass != 0 {
-                    (*(f.frame_thread.pal).offset(
-                        (((t.by >> 1) + (t.bx & 1)) as isize * (f.b4_stride >> 1)
-                            + ((t.bx >> 1) + (t.by & 1)) as isize) as isize,
-                    ))
-                    .as_mut_ptr()
+            let mut set_ctx = |dir: &mut BlockContext, _diridx, off, mul, rep_macro: SetCtxFn| {
+                rep_macro(dir.uvmode.as_mut_ptr(), off, mul * b.uv_mode() as u64);
+            };
+            case_set(cbh4, &mut t.l, 1, cby4 as isize, &mut set_ctx);
+            case_set(cbw4, &mut *t.a, 0, cbx4 as isize, &mut set_ctx);
+
+            if b.pal_sz()[1] != 0 {
+                let pal = if t.frame_thread.pass != 0 {
+                    let index = ((t.by >> 1) + (t.bx & 1)) as isize * (f.b4_stride >> 1)
+                        + ((t.bx >> 1) + (t.by & 1)) as isize;
+                    &*f.frame_thread.pal.offset(index)
                 } else {
-                    (t.scratch.c2rust_unnamed_0.pal).as_mut_ptr()
-                }) as *const [uint16_t; 8];
-                let mut pl = 1;
-                while pl <= 2 {
-                    let mut x_2 = 0;
-                    while x_2 < bw4 {
-                        memcpy(
-                            ((*t).al_pal[0][(bx4 + x_2) as usize][pl as usize]).as_mut_ptr()
-                                as *mut libc::c_void,
-                            (*pal_0.offset(pl as isize)).as_ptr() as *const libc::c_void,
-                            16 as libc::c_int as libc::c_ulong,
-                        );
-                        x_2 += 1;
+                    &t.scratch.c2rust_unnamed_0.pal
+                };
+
+                for pl in 0..2 {
```

Comment:
```suggestion
                for pl in 1..=2 {
```

This comes from this:

```rust
                let mut pl = 1;
                while pl <= 2 {
```

Reply:
Oh yup, you're absolutely right.